pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -282,12 +313,20 @@ func (s *srv) handleTraffic(c echo.Context) error {
 				logger.WithError(err).Warn("Failed to unmarshal Tx acknowledgement frame")
 				return nil
 			}
-			//TODO: Add Token check after rebasing https://github.com/TheThingsNetwork/lorawan-stack/pull/589
-			txAck := messages.ToTxAcknowledgment(nil)
-			if err := conn.HandleTxAck(&txAck); err != nil {
-				logger.WithError(err).Warn("Failed to handle Tx acknowledgement message")
+			var refTime time.Time
+			sec, nsec := math.Modf(txConf.RefTime)
+			if sec != 0 {
+				refTime = time.Unix(int64(sec), int64(nsec*(1e9)))
+			}
+			if correlationIDs, timeDiff, ok := s.tokens.Get(uint16(txConf.Diid), refTime); ok {
```

Comment:
`cids, delta, ok`

Reply:
What I think they mean in [the docs](https://doc.sm.tc/station/tcproto.html#monitoring-round-trip-times) is that we need to subtract `RefTime` from our server time to get the RTT.

So, `rtt = receivedAt.Subtract(refTime)`

Until we send a new `MuxTime`, that `RefTime` keeps increasing from the last `MuxTime` and it represents the server time of what the gateway thinks we are. It will be slightly behind because of network latency, so that's our RTT.

I don't think we need the `io.DownlinkTokens` for timing here; only for correlation IDs. For timing, it is designed to work with two server times (sent and receive), while BS gives you the server time already.

Reply:
Ah yes you're right indeed and the source code agrees (the docs are not very clear though):
```
reftime = s2ctx->muxtime + ts_normalizeTimespanMCU(rt_getTime()-s2ctx->reftime) / 1e6;
(reftime = muxtime + processing time)
``` 
The idea for using only TxAck was that I needed to reference them to the token. Now I can extend this to all uplinks. 👍